packages/angular/cli/lib/init.ts

```diff
@@ -82,7 +82,7 @@ let forceExit = false;
       // If using the update command and the global version is greater, use the newer update command
       // This allows improvements in update to be used in older versions that do not have bootstrapping
       if (
-        process.argv[2] === 'update' &&
+        (process.argv[2] === 'update' || process.argv[2] === 'completion') &&
```

Comment:
While my suggested change was indeed to the logic in here. This change in logic isn't quite right.

This would cause the global CLI to be used if major version is <=1 from the local version. n general we should always use the local CLI, update is an edge case where we do this to avoid downloading a temporary Angular CLI to run the update. 

We should skip the entire `isGlobalGreater` when the command name is `completion`.

```diff
-    let isGlobalGreater = false;
     try {
-      isGlobalGreater = !!localVersion && globalVersion.compare(localVersion) > 0;
-    } catch (error) {
-      // eslint-disable-next-line  no-console
-      console.error('Version mismatch check skipped. Unable to compare local version: ' + error);
-    }
+      const isGlobalGreater = !!localVersion && globalVersion.compare(localVersion) > 0;
+      const rawCommandName = process.argv[2];
 
-    if (isGlobalGreater) {
-      // If using the update command and the global version is greater, use the newer update command
-      // This allows improvements in update to be used in older versions that do not have bootstrapping
-      if (
-        process.argv[2] === 'update' &&
-        cli.VERSION &&
-        cli.VERSION.major - globalVersion.major <= 1
-      ) {
-        cli = await import('./cli');
-      } else if (await isWarningEnabled('versionMismatch')) {
-        // Otherwise, use local version and warn if global is newer than local
-        const warning =
-          `Your global Angular CLI version (${globalVersion}) is greater than your local ` +
-          `version (${localVersion}). The local Angular CLI version is used.\n\n` +
-          'To disable this warning use "ng config -g cli.warnings.versionMismatch false".';
+      // When using the completion command and the global version is greater,
+      // don't show the warning as otherwise this will break completion.
+      if (isGlobalGreater && rawCommandName !== 'completion') {
+        // If using the update command and the global version is greater, use the newer update command
+        // This allows improvements in update to be used in older versions that do not have bootstrapping.
+        if (
+          rawCommandName === 'update' &&
+          cli.VERSION &&
+          cli.VERSION.major - globalVersion.major <= 1
+        ) {
+          cli = await import('./cli');
+        } else if (await isWarningEnabled('versionMismatch')) {
+          // Otherwise, use local version and warn if global is newer than local
+          const warning =
+            `Your global Angular CLI version (${globalVersion}) is greater than your local ` +
+            `version (${localVersion}). The local Angular CLI version is used.\n\n` +
+            'To disable this warning use "ng config -g cli.warnings.versionMismatch false".';
 
-        // eslint-disable-next-line  no-console
-        console.error(colors.yellow(warning));
+          // eslint-disable-next-line  no-console
+          console.error(colors.yellow(warning));
+        }
       }
+    } catch (error) {
+      // eslint-disable-next-line  no-console
+      console.error('Version mismatch check skipped. Unable to compare local version: ' + error);
     }
```

Reply:
Thanks Alan, TBH I'm still a bit lost in the logic for this file and didn't get a chance to look into it too much today to really understand your solution. Would it be easier for you to just take it over given that you already have a full patch there and clearly understand it better than I do?

Reply:
Sure, I can take over.